`gh pr checkout` panics if the PR repo does not exist

### Describe the bug

Upon trying to checkout to a pr with `gh pr checkout 2556` in `Chatterino/chatterino2` you'll get a panic and gh cli will crash because for that specific Pull Request fork on which PR branch was on got deleted.

Version: `gh version 1.10.3 (2021-05-22)`

### Steps to reproduce the behavior

1. Go to cloned `Chatterino/chatterino2` repo.
2. Type `gh pr checkout 2556`
3. See gh cli panic and crash

### Expected vs actual behavior

gh cli should check out to the PR's branch, similar to how the following command would do it: `git fetch origin pull/$pr/head:pr/$pr && git checkout pr/$pr` (where $pr is the PR number, e.g. 2556 in this scenario).
As of writing the issue, gh cli simply panics.

### Logs

```
zneix@uds ~/g/chatterino2 (master)$ gh --version
gh version 1.10.3 (2021-05-22)
https://github.com/cli/cli/releases/tag/v1.10.3
zneix@uds ~/g/chatterino2 (master)$ gh pr checkout 2556
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5586b3af1164]

goroutine 1 [running]:
github.com/cli/cli/pkg/cmd/pr/checkout.checkoutRun(0xc000146af0, 0xc000497680, 0x0)
        github.com/cli/cli/pkg/cmd/pr/checkout/checkout.go:100 +0xac4
github.com/cli/cli/pkg/cmd/pr/checkout.NewCmdCheckout.func1(0xc0003cc000, 0xc0008154d0, 0x1, 0x1, 0x0, 0x0)
        github.com/cli/cli/pkg/cmd/pr/checkout/checkout.go:61 +0x166
github.com/spf13/cobra.(*Command).execute(0xc0003cc000, 0xc0008154b0, 0x1, 0x1, 0xc0003cc000, 0xc0008154b0)
        github.com/spf13/cobra@v1.1.3/command.go:852 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000b3400, 0xc0008153a0, 0x5586b3bb7501, 0xd)
        github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
main.mainRun(0xc00002e0b8)
        github.com/cli/cli/cmd/gh/main.go:166 +0x676
main.main()
        github.com/cli/cli/cmd/gh/main.go:46 +0x25
zneix@uds ~/g/chatterino2 (master) [2]$ git remote -v
fork    gh:/zneix/chatterino2 (fetch)
fork    gh:/zneix/chatterino2 (push)
origin  gh:/Chatterino/chatterino2 (fetch)
origin  gh:/Chatterino/chatterino2 (push)
zneix@uds ~/g/chatterino2 (master)$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
zneix@uds ~/g/chatterino2 (master)$
```

--- a/pkg/cmd/pr/checkout/checkout.go
+++ b/pkg/cmd/pr/checkout/checkout.go
@@ -72,7 +72,7 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
 func checkoutRun(opts *CheckoutOptions) error {
 	findOptions := shared.FindOptions{
 		Selector: opts.SelectorArg,
-		Fields:   []string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository"},
+		Fields:   []string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository", "maintainerCanModify"},
 	}
 	pr, baseRepo, err := opts.Finder.Find(findOptions)
 	if err != nil {

--- a/pkg/cmd/pr/checkout/checkout.go
+++ b/pkg/cmd/pr/checkout/checkout.go
@@ -96,7 +96,9 @@ func checkoutRun(opts *CheckoutOptions) error {
 	}
 
 	headRemote := baseRemote
-	if pr.IsCrossRepository {
+	if pr.HeadRepository == nil {
+		headRemote = nil
+	} else if pr.IsCrossRepository {
 		headRemote, _ = remotes.FindByRepo(pr.HeadRepositoryOwner.Login, pr.HeadRepository.Name)
 	}
 
@@ -207,7 +209,7 @@ func cmdsForMissingRemote(pr *api.PullRequest, baseURLOrName, repoHost, defaultB
 
 	remote := baseURLOrName
 	mergeRef := ref
-	if pr.MaintainerCanModify {
+	if pr.MaintainerCanModify && pr.HeadRepository != nil {
 		headRepo := ghrepo.NewWithHost(pr.HeadRepositoryOwner.Login, pr.HeadRepository.Name, repoHost)
 		remote = ghrepo.FormatRemoteURL(headRepo, protocol)
 		mergeRef = fmt.Sprintf("refs/heads/%s", pr.HeadRefName)

--- a/pkg/cmd/pr/checkout/checkout_test.go
+++ b/pkg/cmd/pr/checkout/checkout_test.go
@@ -2,6 +2,7 @@ package checkout
 
 import (
 	"bytes"
+	"errors"
 	"io/ioutil"
 	"net/http"
 	"strings"
@@ -58,6 +59,99 @@ func stubPR(repo, prHead string) (ghrepo.Interface, *api.PullRequest) {
 		MaintainerCanModify: false,
 	}
 }
+
+func Test_checkoutRun(t *testing.T) {
+	tests := []struct {
+		name       string
+		opts       *CheckoutOptions
+		httpStubs  func(*httpmock.Registry)
+		runStubs   func(*run.CommandStubber)
+		remotes    map[string]string
+		wantStdout string
+		wantStderr string
+		wantErr    bool
+	}{
+		{
+			name: "fork repo was deleted",
+			opts: &CheckoutOptions{
+				SelectorArg: "123",
+				Finder: func() shared.PRFinder {
+					baseRepo, pr := stubPR("OWNER/REPO:master", "hubot/REPO:feature")
+					pr.MaintainerCanModify = true
+					pr.HeadRepository = nil
+					finder := shared.NewMockFinder("123", pr, baseRepo)
+					return finder
+				}(),
+				Config: func() (config.Config, error) {
+					return config.NewBlankConfig(), nil
+				},
+				Branch: func() (string, error) {
+					return "main", nil
+				},
+			},
+			remotes: map[string]string{
+				"origin": "OWNER/REPO",
+			},
+			runStubs: func(cs *run.CommandStubber) {
+				cs.Register(`git fetch origin refs/pull/123/head:feature`, 0, "")
+				cs.Register(`git config branch\.feature\.merge`, 1, "")
+				cs.Register(`git checkout feature`, 0, "")
+				cs.Register(`git config branch\.feature\.remote origin`, 0, "")
+				cs.Register(`git config branch\.feature\.merge refs/pull/123/head`, 0, "")
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			opts := tt.opts
+
+			io, _, stdout, stderr := iostreams.Test()
+			opts.IO = io
+
+			httpReg := &httpmock.Registry{}
+			defer httpReg.Verify(t)
+			if tt.httpStubs != nil {
+				tt.httpStubs(httpReg)
+			}
+			opts.HttpClient = func() (*http.Client, error) {
+				return &http.Client{Transport: httpReg}, nil
+			}
+
+			cmdStubs, cmdTeardown := run.Stub()
+			defer cmdTeardown(t)
+			if tt.runStubs != nil {
+				tt.runStubs(cmdStubs)
+			}
+
+			opts.Remotes = func() (context.Remotes, error) {
+				if len(tt.remotes) == 0 {
+					return nil, errors.New("no remotes")
+				}
+				var remotes context.Remotes
+				for name, repo := range tt.remotes {
+					r, err := ghrepo.FromFullName(repo)
+					if err != nil {
+						return remotes, err
+					}
+					remotes = append(remotes, &context.Remote{
+						Remote: &git.Remote{Name: name},
+						Repo:   r,
+					})
+				}
+				return remotes, nil
+			}
+
+			err := checkoutRun(opts)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("want error: %v, got: %v", tt.wantErr, err)
+			}
+			assert.Equal(t, tt.wantStdout, stdout.String())
+			assert.Equal(t, tt.wantStderr, stderr.String())
+		})
+	}
+}
+
+/** LEGACY TESTS **/
 
 func runCommand(rt http.RoundTripper, remotes context.Remotes, branch string, cli string) (*test.CmdOut, error) {
 	io, _, stdout, stderr := iostreams.Test()

--- a/pkg/cmd/pr/checkout/checkout_test.go
+++ b/pkg/cmd/pr/checkout/checkout_test.go
@@ -205,7 +205,7 @@ func TestPRCheckout_sameRepo(t *testing.T) {
 
 	baseRepo, pr := stubPR("OWNER/REPO", "OWNER/REPO:feature")
 	finder := shared.RunCommandFinder("123", pr, baseRepo)
-	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository"})
+	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository", "maintainerCanModify"})
 
 	cs, cmdTeardown := run.Stub()
 	defer cmdTeardown(t)
@@ -260,7 +260,7 @@ func TestPRCheckout_differentRepo_remoteExists(t *testing.T) {
 
 	baseRepo, pr := stubPR("OWNER/REPO", "hubot/REPO:feature")
 	finder := shared.RunCommandFinder("123", pr, baseRepo)
-	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository"})
+	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository", "maintainerCanModify"})
 
 	cs, cmdTeardown := run.Stub()
 	defer cmdTeardown(t)
@@ -283,7 +283,7 @@ func TestPRCheckout_differentRepo(t *testing.T) {
 
 	baseRepo, pr := stubPR("OWNER/REPO:master", "hubot/REPO:feature")
 	finder := shared.RunCommandFinder("123", pr, baseRepo)
-	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository"})
+	finder.ExpectFields([]string{"number", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository", "maintainerCanModify"})
 
 	cs, cmdTeardown := run.Stub()
 	defer cmdTeardown(t)

